Parallel.pm

```diff
@@ -0,0 +1,195 @@
+    $self->{PROC}    = $proc;
+    $self->{COUNT}   = ( $count   ) ? $count : $COUNT_1;
+    $self->{TIMEOUT} = ( $timeout ) ? $timeout : $MIN_60 * $MIN_60; # 1 hour
+
+    $self->{CHILDREN} = {};
+    $self->{CHILDS_ERROR} = {};
+
+    bless( $self, $class );
+    return $self;
+}
+
+######################################################################
+#
+#  Log
+#
+sub Log {
+    my $str = shift;
+    print STDOUT "$str\n";
+
+    return 0;
+}
+
+######################################################################
+# parallel_run
+# Return: 0 - ok
+#         1 - error
+#
+sub parallel_run {
+    my $self = shift;
+    my $debug = $self->{DEBUG};
+    my $date;
+    my $children;
+    my $childs_error;
+
+    local $SIG{CHLD} = sub {
+        my $children;
+        my $childs_error;
+        $children = $self->{CHILDREN};
+        $childs_error = $self->{CHILDS_ERROR};
+
+        # don't change $! and $? outside handler
+        local ($ERRNO, $CHILD_ERROR) = ($ERRNO, $CHILD_ERROR);
+        my $child_pid = waitpid $NUMBER_NEGATIVE_1, WNOHANG;
+        if ( $child_pid == $NUMBER_NEGATIVE_1 or $child_pid == $NUMBER_0 ) { return; }
+        $date = strftime '%A, %d %B %Y %H:%M:%S GMT', localtime;
+        Log( "[$date] Parent: child $child_pid was reaped" );
+        if ( ! defined $childs_error->{$child_pid} ) { return; }
+        $childs_error->{$child_pid} = $CHILD_ERROR;
+        if ( $debug ) { Log( "[DEBUG] SIG child $child_pid, rc=$CHILD_ERROR" ); }
+        if ( defined $children->{$child_pid} ) { delete $children->{$child_pid}; }
+        if ( $childs_error->{$child_pid} > 0 ) {
+            $self->{RC} = 1;
+            if ( $debug ) { Log( "[DEBUG] SIG child error for $child_pid, rc=$CHILD_ERROR" ); }
+        }
+    };
+
+    eval {
+        local $SIG{ALRM} = sub { $self->{RC} = 1; croak 'alarm timeout'; };
+        alarm $self->{TIMEOUT};
+        $self->parallel_start();
+        alarm 0;
+    };
+
+    Log ( "[INFO] parallel run finished at " . ( strftime '%A, %d %B %Y %H:%M:%S GMT', localtime ) );
+    if ( $EVAL_ERROR && $EVAL_ERROR !~ quotemeta 'alarm timeout' ) {
+        $self->{RC} = 1; croak "[ERROR] $EVAL_ERROR";
+    } else {
+        foreach my $child_pid ( sort keys %$childs_error ) {
+            if ( $childs_error->{$child_pid} > 0 ) {
+                if ( $debug ) { Log( "[DEBUG] Child error for $child_pid, rc=" . $childs_error->{$child_pid} ); }
+                $self->{RC} = 1;
+                last;
+            }
+        }
+    }
+
+    return $self->{RC};
+}
+
+######################################################################
+# parallel_start
+#
+sub parallel_start {
+    my $self = shift;
+    my $num = $self->{COUNT};
+    my @counts = ( $COUNT_1..$num );
+    my $start_time = time();
+    my $end_time;
+    my $tout;
+    my $date;
+    my $children;
+    my $childs_error;
+
+    $children = $self->{CHILDREN};
+    $childs_error = $self->{CHILDS_ERROR};
+
+    foreach my $count (@counts) {
```

Comment:
There is no requirement to run `$USER_COMMAND` multiple times per host.
So, the implementation is overcomplicated.